src/classes/LVL_LevelAssign_BATCH.cls

```diff
@@ -267,35 +284,149 @@ public without sharing class LVL_LevelAssign_BATCH implements Database.Batchable
+        */
+        public Ladder(String target) {
+            this.target = target;
+        }
+
+        /*********************************************************************************************************
+        * @description Add a level to the ladder, caching query fields, source fields and the most recent
+        * LastModifiedDate of all the levels in the ladder
+        */
+        public Ladder addLevel(Level__c level) {
+            addQueryFields(level);
+
+            sourceFields.add(level.Source_Field__c);
+
+            if (mostRecentLastModifiedDate == null || mostRecentLastModifiedDate < level.LastModifiedDate) {
+                mostRecentLastModifiedDate = level.LastModifiedDate;
+            }
+
+            return this;
+        }
+
+        /*********************************************************************************************************
+        * @description Builds query to get object data need to calculate and assign levels
+        * @returns SOQL string for querying SObjects for level calculation
+        */
+        public String buildQuery() {
+            String soqlQuery = 'SELECT Id';
+            for (String fieldName : queryFields) {
+                soqlQuery += ', ' + fieldName;
+            }
+            soqlQuery += ' FROM ' + target + buildFilter();
+            return soqlQuery;
+        }
+
+        /*********************************************************************************************************
+        * @description Adds fields from Level that need to be queried
+        */
+        private void addQueryFields(Level__c level) {
+            queryFields.add(level.Source_Field__c);
+            queryFields.add(level.Level_Field__c);
+            if (!String.isEmpty(level.Previous_Level_Field__c)) {
+                queryFields.add(level.Previous_Level_Field__c);
             }
-            setFlds.add(lvl.Source_Field__c);
-            setFlds.add(lvl.Level_Field__c);
-            if (lvl.Previous_Level_Field__c != null)
-                setFlds.add(lvl.Previous_Level_Field__c);            
         }
         
-        return mapStrObjectToQueryFields; 
+        /*********************************************************************************************************
+        * @description Builds query to calulate level for the target object with filter if necessary
+        * @returns SOQL string
+        */
+        private String buildFilter() {
+            if (!shouldBuildFilter()) {
+                return '';
+            }
+
+            DateTime datetimeToFilter;
+            if (targetIsAccount) {
+                datetimeToFilter = levelsSettings.LastJobStartTimeAccount__c;
+            } else if (targetIsContact) {
+                datetimeToFilter = levelsSettings.LastJobStartTimeContact__c;
+            } else {
+                return '';
+            }
+
+            return ' WHERE SystemModStamp > ' + datetimeToFilter.formatGMT(DATETIME_SOQL_FORMAT);
```

Comment:
Do you think this should be `>=` to catch Accounts/Contacts that might have been updated at exactly the same moment the job previously started?

Reply:
That's probably safest the safest approach, I'll do that